File: python/23.02/33.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1:
            return 0 if target == nums[0] else -1

        s, e = 0, len(nums)
        i, j = 0, len(nums)
        x = 0
        while x < 10:
            mid = (i+j) // 2
            if nums[0] <= nums[mid-1] and nums[0] >= nums[mid]:
                if target < nums[0] or mid == 0:
                    s, e = mid, len(nums)
                else:
                    s, e = 0, mid
                break
            if nums[0] < nums[mid]:
                i = mid
            else:
                j = mid + 1
            x += 1

        if target == nums[s]:
            return s

        while s < e:
            mid = (s+e) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                s = mid + 1
            else:
                e = mid
        
        return -1

File: typescript/23.09/33.ts
function search(nums: number[], target: number): number {
    if (nums.length === 1) return target === nums[0] ? 0 : -1

    let [s, e] = [0, nums.length]
    let [i, j] = [0, nums.length]
    let x = 0
    
    while (x < 10) {
        let mid = Math.floor((i+j)/2)
        if (nums[0] <= nums[mid-1] && nums[0] >= nums[mid]) {
            if (target < nums[0] || mid === 0) {
                s = mid
                e = nums.length
            } else {
                s = 0
                e = mid
            }
            break
        }
            
        if (nums[0] < nums[mid]) {
            i = mid
        } else {
            j = mid + 1
        }
        x++
    }

    if (target === nums[s]) return s

    while (s < e) {
        let mid = Math.floor((s+e)/2)
        if (nums[mid] == target) return mid
        if (nums[mid] < target) {
            s = mid + 1
        } else {
            e = mid
        }
    }
    
    return -1
};
